src/vos/vos_obj_cache.c

```diff
@@ -320,12 +435,16 @@ vos_obj_hold(struct daos_lru_cache *occ, struct vos_container *cont,
 	if (cont->vc_pool->vp_dying)
```

Comment:
To be honest, it's bit hard for me to follow this vos_obj_hold() function already :) , but seems we have to complicate it further (in this PR and in md-on-ssd phase 2), so I'm wondering if we'd split this gigantic vos_obj_hold() for the sake of code maintainance?

It looks to me there are three major functionalities implemented in current vos_obj_hold():
1. Object cache lookup/create;
2. OI lookup/create;
3. ilog check;

Is it possible to make it into three individual functions respectively?

I think the pure object cache hold function (without those 'intent' and ilog complexities involved) would be very helpful for future extension. (In md-on-ssd phase 2, we will need a function to hold object cache & pin the necessary memory buckets, and it could be called in vos_update_begin() no matter if the OI exists or not.)

Does this proposal sound good to you? Anything I overlooked?

Reply:
I don't disagree with it being difficult to follow

Are you suggesting callers would have 3 functions to call?  Not necessarily opposed to it but I don't think this particular patch is the place to do it.

I might need you to elaborate on the phase 2 plan a bit.   Why does memory need to be pinned in vos_update_begin?  Maybe we can chat about it in the MD on SSD call.

Reply:
In phase 2, we'll have to know which bucket the object is using (or will use when the object doesn't exist) in vos_update_begin(), and the bucket has to be pinned in memory until vos_update_end(), few allocations (reserve) could happen between vos_update_begin() and vos_update_end():
1. small value located on meta blob;
2. single value vos_irec_df & csum;
 
To ensure concurrent updates use same bucket for reserve, I think the best place to track bucket info would be object cache, and since bucket pin/load operation could yield, I'm afraid that we might need to make vos_obj_hold() into at least two parts in the future:
1. Pure object cache part, which includes bucket pin and being called outside of tx.
2. Other part which might need be called within tx. 

Yeah, let's discuss it in the meeting.